pkg/workspace/library_loader.go

```diff
@@ -0,0 +1,190 @@
+package workspace
+
+import (
+	"fmt"
+
+	"github.com/k14s/ytt/pkg/files"
+	"github.com/k14s/ytt/pkg/texttemplate"
+	"github.com/k14s/ytt/pkg/yamlmeta"
+	"github.com/k14s/ytt/pkg/yamltemplate"
+	"github.com/k14s/ytt/pkg/yttlibrary"
+)
+
+type LibraryLoader struct {
+	*Library
+	UI     files.UI
+	Opts   TemplateLoaderOpts
+	Values interface{}
+}
+
+type EvalResult struct {
+	DocSet      *yamlmeta.DocumentSet
+	DocSets     map[*FileInLibrary]*yamlmeta.DocumentSet
+	OutputFiles []files.OutputFile
+}
+
+func NewLibraryLoader(lib *Library, parentValuesAst interface{}, ui files.UI, loaderOpts TemplateLoaderOpts) *LibraryLoader {
+	return &LibraryLoader{
+		UI:      ui,
+		Opts:    loaderOpts,
+		Library: lib,
+		Values:  parentValuesAst,
+	}
+}
+
+func (ll *LibraryLoader) LoadValues() error {
+	valuesFiles, err := ll.loadValueFiles(ll.Library, true)
+	if err != nil {
+		return err
+	}
+
+	noValuesLoader := NewTemplateLoader(nil, ll.UI, ll.Opts)
+
+	ll.Values, err = DataValuesPreProcessing{valuesFiles, ll.Values, noValuesLoader, ll.Opts.IgnoreUnknownComments}.Apply()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ll *LibraryLoader) loadValueFiles(lib *Library, recurse bool) ([]*FileInLibrary, error) {
+	var valuesFiles []*FileInLibrary
+
+	for _, file := range lib.files {
+		fileInLib := &FileInLibrary{file, lib}
+
+		if fileInLib.File.Type() == files.TypeYAML && fileInLib.File.IsTemplate() {
+			fileBs, err := fileInLib.File.Bytes()
+			if err != nil {
+				return nil, err
+			}
+
+			docSet, err := yamlmeta.NewDocumentSetFromBytes(fileBs, fileInLib.File.RelativePath())
+			if err != nil {
+				return nil, fmt.Errorf("Unmarshaling YAML template '%s': %s", fileInLib.File.RelativePath(), err)
+			}
+
+			tplOpts := yamltemplate.MetasOpts{IgnoreUnknown: ll.Opts.IgnoreUnknownComments}
+
+			valuesDocs, _, err := yttlibrary.DataValues{docSet, tplOpts}.Extract()
+			if err != nil {
+				return nil, err
+			}
+
+			if len(valuesDocs) > 0 {
+				valuesFiles = append(valuesFiles, fileInLib)
+				file.MarkForOutput(false)
+			}
+		}
+	}
+
+	if recurse {
```

Comment:
let's take out recursion out of this PR.

Reply:
I specifically avoided `ListAccessibleFiles()` and implemented recursion here for #31. If I want template evaluation to change the recursion strategy, I need to perform the recursion after template evaluation and not before as `ListAccessibleFiles()` would do.

I can move this code in #31 and keep `ListAccessibleFiles()` here.

Reply:
This is not the global `-R` flag we talked about in #24, but rather a preparation for what I mentioned https://github.com/k14s/ytt/pull/30#discussion_r296390561 and https://github.com/k14s/ytt/pull/24#issuecomment-503953521 where value files would be considered for each directory as they are processed and not once for all before.

This however is forward work that might belong to other PRs, I'm ok with removing it from this PR.